plugin/pkg/auth/authenticator/token/keystone/keystone.go

```diff
@@ -0,0 +1,58 @@
+/*
+Copyright 2015 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package keystone
+
+import (
+	"errors"
+
+	"github.com/golang/glog"
+	"k8s.io/kubernetes/pkg/auth/user"
+)
+
+// KeystoneAuthenticator contacts openstack keystone to validate user's credentials passed in the request.
+// The keystone endpoint is passed during apiserver startup
+type KeystoneAuthenticator struct {
+	configPath       string
+	apiCallValidator validator
+}
+
+func (keystoneAuthenticator *KeystoneAuthenticator) AuthenticateToken(token string) (user.Info, bool, error) {
+	//FIXME kfox1111: To enhance performance, cache admin token/client.
+	glog.V(3).Infof("Starting Keystone Token Auth.")
+	userid, projectid, roles, valid, err := keystoneAuthenticator.apiCallValidator.validate(token)
+	if err != nil {
+		return nil, false, err
+	}
+	extra := make(map[string][]string, 3)
+	extra["alpha.kubernetes.io/keystone_authn"] = []string{"true"}
+	extra["alpha.kubernetes.io/keystone_project_id"] = []string{projectid}
+	extra["alpha.kubernetes.io/keystone_roles"] = roles
```

Comment:
See the main discussion thread on why I think it might be bad to map the keystone concepts to groups. It can be done, but I think there are some potential negative affects from that. I would like to reuse the RBAC api to do the authz stuff, but am not sure if it should be done by adding keystone project/role knowledge into RBAC, or trying to munge the info into groups and ensure no group caching bites us, or that you don't have to start parsing out stuff out of parts of group names.